packages/smart/contracts/turbo/CryptoMarketFactory.sol

```diff
@@ -0,0 +1,219 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.7.6;
+pragma abicoder v2;
+
+import "../libraries/IERC20Full.sol";
+import "../balancer/BPool.sol";
+import "./AbstractMarketFactory.sol";
+import "./FeePot.sol";
+import "../libraries/SafeMathInt256.sol";
+
+import "@chainlink/contracts/src/v0.7/interfaces/AggregatorV3Interface.sol";
+
+contract CryptoMarketFactory is AbstractMarketFactory {
+    using SafeMathUint256 for uint256;
+    using SafeMathInt256 for int256;
+
+    event MarketCreated(
+        uint256 id,
+        address creator,
+        uint256 indexed endTime,
+        MarketType marketType,
+        uint256 indexed coinIndex,
+        uint256 price
+    );
+    event MarketResolved(uint256 id, address winner);
+
+    struct Coin {
+        string name;
+        AggregatorV3Interface priceFeed;
+        uint256 price;
+        uint256[1] currentMarkets;
+    }
+    Coin[] public coins;
+
+    enum MarketType {
+        PriceUpDown // 0
+    }
+    enum PriceUpDownOutcome {
+        Above, // 0
+        NotAbove // 1
+    }
+    struct MarketDetails {
+        MarketType marketType;
+        uint256 coinIndex;
+        uint256 price;
```

Comment:
Is there a resolve timestamp for the market? 
Need to tell the betters the timestamp the price will be used for resolution.